File: app/src/main/java/piuk/blockchain/android/ui/contacts/payments/ContactsPaymentRequestPresenter.java
package piuk.blockchain.android.ui.contacts.payments;

import android.os.Bundle;
import android.support.annotation.VisibleForTesting;

import info.blockchain.wallet.contacts.data.Contact;
import info.blockchain.wallet.contacts.data.PaymentRequest;
import info.blockchain.wallet.contacts.data.RequestForPaymentRequest;

import javax.inject.Inject;

import piuk.blockchain.android.R;
import piuk.blockchain.android.data.contacts.ContactsPredicates;
import piuk.blockchain.android.data.contacts.PaymentRequestType;
import piuk.blockchain.android.data.datamanagers.ContactsDataManager;
import piuk.blockchain.android.data.datamanagers.PayloadDataManager;
import piuk.blockchain.android.ui.base.BasePresenter;
import piuk.blockchain.android.ui.customviews.ToastCustom;

import static piuk.blockchain.android.ui.contacts.payments.ContactPaymentRequestNotesFragment.ARGUMENT_ACCOUNT_POSITION;
import static piuk.blockchain.android.ui.contacts.payments.ContactPaymentRequestNotesFragment.ARGUMENT_CONTACT_ID;
import static piuk.blockchain.android.ui.contacts.payments.ContactPaymentRequestNotesFragment.ARGUMENT_REQUEST_TYPE;
import static piuk.blockchain.android.ui.contacts.payments.ContactPaymentRequestNotesFragment.ARGUMENT_SATOSHIS;


public class ContactsPaymentRequestPresenter extends BasePresenter<ContactPaymentRequestView> {

    private ContactsDataManager contactsDataManager;
    private PayloadDataManager payloadDataManager;
    @VisibleForTesting Contact recipient;
    @VisibleForTesting long satoshis;
    @VisibleForTesting int accountPosition;
    @VisibleForTesting PaymentRequestType paymentRequestType;

    @Inject
    ContactsPaymentRequestPresenter(ContactsDataManager contactsDataManager,
                                    PayloadDataManager payloadDataManager) {

        this.contactsDataManager = contactsDataManager;
        this.payloadDataManager = payloadDataManager;
    }

    @Override
    public void onViewReady() {
        Bundle fragmentBundle = getView().getFragmentBundle();
        String contactId = fragmentBundle.getString(ARGUMENT_CONTACT_ID);
        satoshis = fragmentBundle.getLong(ARGUMENT_SATOSHIS, -1L);
        accountPosition = fragmentBundle.getInt(ARGUMENT_ACCOUNT_POSITION, -1);
        paymentRequestType = (PaymentRequestType) fragmentBundle.getSerializable(ARGUMENT_REQUEST_TYPE);

        if (contactId != null && paymentRequestType != null && satoshis > -1L) {
            loadContact(contactId);
        } else {
            throw new AssertionError("Contact ID and PaymentRequestType must be passed to fragment");
        }
    }

    void sendRequest() {
        if (satoshis <= 0) {
            getView().showToast(R.string.invalid_amount, ToastCustom.TYPE_ERROR);
        } else {
            getView().showProgressDialog();

            if (paymentRequestType.equals(PaymentRequestType.REQUEST)) {

                PaymentRequest paymentRequest = new PaymentRequest(satoshis, getView().getNote());

                getCompositeDisposable().add(
                        payloadDataManager.getNextReceiveAddress(accountPosition)
                                .doOnNext(paymentRequest::setAddress)
                                // Request that the other person sends payment
                                .flatMapCompletable(s -> contactsDataManager.requestSendPayment(recipient.getMdid(), paymentRequest))
                                .doAfterTerminate(() -> getView().dismissProgressDialog())
                                .subscribe(
                                        () -> getView().showSendSuccessfulDialog(recipient.getName()),
                                        throwable -> getView().showToast(R.string.contacts_error_sending_payment_request, ToastCustom.TYPE_ERROR)));

            } else {
                RequestForPaymentRequest request = new RequestForPaymentRequest(satoshis, getView().getNote());
                getCompositeDisposable().add(
                        // Request that the other person receives payment
                        contactsDataManager.requestReceivePayment(recipient.getMdid(), request)
                                .doAfterTerminate(() -> getView().dismissProgressDialog())
                                .subscribe(
                                        () -> getView().showRequestSuccessfulDialog(),
                                        throwable -> getView().showToast(R.string.contacts_error_sending_payment_request, ToastCustom.TYPE_ERROR)));
            }
        }
    }

    private void loadContact(String contactId) {
        getCompositeDisposable().add(
                contactsDataManager.getContactList()
                        .filter(ContactsPredicates.filterById(contactId))
                        .subscribe(
                                contact -> {
                                    recipient = contact;
                                    getView().contactLoaded(recipient.getName(), paymentRequestType);
                                },
                                throwable -> {
                                    getView().showToast(R.string.contacts_not_found_error, ToastCustom.TYPE_ERROR);
                                    getView().finishPage();
                                },
                                () -> {
                                    if (recipient == null) {
                                        // Wasn't found via filter, show not found
                                        getView().showToast(R.string.contacts_not_found_error, ToastCustom.TYPE_ERROR);
                                        getView().finishPage();
                                    }
                                }));
    }

}

File: app/src/main/java/piuk/blockchain/android/ui/contacts/payments/ContactsPaymentRequestPresenter.kt
package piuk.blockchain.android.ui.contacts.payments

import android.support.annotation.VisibleForTesting
import info.blockchain.wallet.contacts.data.Contact
import info.blockchain.wallet.contacts.data.RequestForPaymentRequest
import piuk.blockchain.android.R
import piuk.blockchain.android.data.contacts.ContactsDataManager
import piuk.blockchain.android.data.contacts.ContactsPredicates
import piuk.blockchain.android.data.rxjava.RxUtil
import piuk.blockchain.android.ui.account.PaymentConfirmationDetails
import piuk.blockchain.android.ui.base.BasePresenter
import piuk.blockchain.android.ui.contacts.payments.ContactPaymentRequestNotesFragment.Companion.ARGUMENT_CONFIRMATION_DETAILS
import piuk.blockchain.android.ui.contacts.payments.ContactPaymentRequestNotesFragment.Companion.ARGUMENT_CONTACT_ID
import piuk.blockchain.android.ui.contacts.payments.ContactPaymentRequestNotesFragment.Companion.ARGUMENT_SATOSHIS
import piuk.blockchain.android.ui.customviews.ToastCustom
import javax.inject.Inject

class ContactsPaymentRequestPresenter @Inject internal constructor(
        private val contactsDataManager: ContactsDataManager
) : BasePresenter<ContactPaymentRequestView>() {

    @VisibleForTesting internal var recipient: Contact? = null
    @VisibleForTesting internal var satoshis: Int? = null

    override fun onViewReady() {
        val fragmentBundle = view.fragmentBundle
        val contactId = fragmentBundle.getString(ARGUMENT_CONTACT_ID)
        val confirmationDetails: PaymentConfirmationDetails =
                fragmentBundle.getParcelable(ARGUMENT_CONFIRMATION_DETAILS)
        satoshis = fragmentBundle.getInt(ARGUMENT_SATOSHIS)

        @Suppress("SENSELESS_COMPARISON")
        if (contactId != null && confirmationDetails != null && satoshis != null) {
            loadContact(contactId)
            updateUi(confirmationDetails)
        } else {
            throw IllegalArgumentException("Contact ID, confirmation details and satoshi amount must be passed to fragment")
        }
    }

    internal fun sendRequest() {
        view.showProgressDialog()
        val request = RequestForPaymentRequest(satoshis!!.toLong(), view.note)
        // Request that the other person receives payment
        contactsDataManager.requestReceivePayment(recipient!!.mdid, request)
                .compose(RxUtil.addCompletableToCompositeDisposable(this))
                .doAfterTerminate { view.dismissProgressDialog() }
                .subscribe(
                        { view.showRequestSuccessfulDialog() },
                        {
                            view.showToast(
                                    R.string.contacts_error_sending_payment_request,
                                    ToastCustom.TYPE_ERROR
                            )
                        })
    }

    private fun updateUi(confirmationDetails: PaymentConfirmationDetails) {
        view.updateAccountName(confirmationDetails.fromLabel)
        view.updateTotalBtc("${confirmationDetails.btcTotal} ${confirmationDetails.btcUnit}")
        view.updateTotalFiat(confirmationDetails.fiatSymbol + confirmationDetails.fiatTotal)
    }

    private fun loadContact(contactId: String) {
        contactsDataManager.getContactList()
                .compose(RxUtil.addObservableToCompositeDisposable(this))
                .filter(ContactsPredicates.filterById(contactId))
                .subscribe(
                        { contact ->
                            recipient = contact
                            view.contactLoaded(recipient!!.name)
                        },
                        {
                            view.showToast(R.string.contacts_not_found_error, ToastCustom.TYPE_ERROR)
                            view.finishPage()
                        },
                        {
                            if (recipient == null) {
                                // Wasn't found via filter, show not found
                                view.showToast(R.string.contacts_not_found_error, ToastCustom.TYPE_ERROR)
                                view.finishPage()
                            }
                        })
    }

}
